bluefog/common/topology_util.py

```diff
@@ -351,5 +425,160 @@ def GetDynamicSendRecvRanks(topo: nx.DiGraph, self_rank: int) -> Iterator[Tuple[
+            source_rank_id = source_machine_id * nodes_per_machine + local_rank_id
+            recv_ranks.append(source_rank_id)
+
+        else:
+            if nodes_per_machine == 2:
+                # Do not send. But our neighbor_allreduce haven't supported this yet.
+                raise ValueError(
+                    "Unfortunately, nodes_per_machine is 2 case is not supported yet.")
+            else:
+                # find send_rank
+                target_local_rank_id = (local_rank_id + 1) % nodes_per_machine
+                if target_local_rank_id == local_rank_to_go_outside_id:
+                    target_local_rank_id = (
+                        target_local_rank_id + 1) % nodes_per_machine
+                target_rank_id = target_local_rank_id + machine_id * nodes_per_machine
+                send_rank = target_rank_id
+
+                # find recv_rank
+                source_local_rank_id = (local_rank_id - 1) % nodes_per_machine
+                if source_local_rank_id == local_rank_to_go_outside_id:
+                    source_local_rank_id = (
+                        source_local_rank_id - 1) % nodes_per_machine
+                source_rank_id = source_local_rank_id + machine_id * nodes_per_machine
+                recv_ranks.append(source_rank_id)
+
+        yield [send_rank], recv_ranks
+        index += 1
+
+
+def GetInnerOuterExp2DynamicSendRecvRanks(
+        world_size: int, local_size: int, self_rank: int
+    ) -> Iterator[Tuple[List[int], List[int]]]:
+    """A utility function to generate 1-outgoing send rank and corresponding recieving rank(s)
+       for Inner-Exp2-Outer-Exp2 ring topology
+
+    Args:
+        world_size (int): the size of all nodes; world_size = num_machines * nodes_per_machine
+        local_size (int): number of nodes in each machine
+        self_rank (int): The self rank.
+
+    Yields:
+        Iterator[Tuple[List[int], List[int]]]: send_ranks, recv_ranks.
+
+    Example:
+
+        >>> from bluefog.common import topology_util
+        >>> world_size, local_size = bf.size(), bf.local_size()
+        >>> gen = topology_util.GetInnerOuterExp2DynamicSendRecvRanks(world_size, local_size, 0)
+        >>> for _ in range(10):
+        >>>     print(next(gen))
+    """
+    num_machines = world_size//local_size
+    nodes_per_machine = local_size
+    assert world_size % local_size == 0, "It should be used under homogeneous environment only."
+    assert nodes_per_machine > 2, "We do not support the case nodes per machine < 2 yet."
+    exp_2_out_size = int(np.log2(num_machines-1))
+    exp_2_in_size = int(np.log2(nodes_per_machine-2))
+
+    index = 0
+    while True:
+        machine_id = self_rank // nodes_per_machine
+        local_rank_id = self_rank % nodes_per_machine
+        local_rank_to_go_outside_id = index % nodes_per_machine
+
+        recv_ranks = []
+        if local_rank_to_go_outside_id == local_rank_id:
+            next_machine_dist = 2**(index % (exp_2_out_size+1))
```

Comment:
If exp2_out_size+1==nodes_per_machine, it means next_machine_dist will be a constant whenever reaches this point. Is this correct?

Reply:
that is right. Under two nodes/machines case, it is always 0 -> 1 and 1->0.

Reply:
Assume `num_machines=16`, `nodes_per_machine=4`, and `self_rank=1`, then we know that `exp_2_out_size=3`, and `local_rank_id=1`. If this branch is reached, `local_rank_to_go_outside_id=1`, and `index % (exp_2_out_size+1)=1`, resulting in `next_machine_dist` always equal to 2. However, I feel for self_rank=1, its machine id is 0, which should communicate with machine 1, 2, 4, 8 in this case. However, here this next_machine_dist never gets value like 1, 4, or 8 here.

Reply:
You are right. That is a design problem. It is not wrong. But maybe we can find a better mixing strategy in the future. Add a note there.